cf/src/main/scala/typeformation/iam/Condition.scala

```diff
@@ -0,0 +1,58 @@
+package typeformation.iam
+
+import io.circe.Encoder
+
+trait Condition {
+  type Type
+  def key: Condition.Key
+  def expected: Type
+  def quantifier: Option[Condition.Quantifier]
+  def hasIfExists: Boolean
+  def label: String
+  def expressionEncoder: Encoder[Type]
+
+  def ifExists: Condition =
+    Condition.instance[Type](quantifier, ifX = true)(key, expected)(label)(
+      expressionEncoder)
+
+  def forAllValues: Condition =
+    Condition.instance[Type](Some(Condition.ForAllValues), hasIfExists)(
+      key,
+      expected)(label)(expressionEncoder)
+
+  def forAnyValues: Condition =
+    Condition.instance[Type](Some(Condition.ForAnyValue), hasIfExists)(
+      key,
+      expected)(label)(expressionEncoder)
+}
+
+object Condition {
+  case class Key(value: String)
+
+  type Aux[A] = Condition {
```

Comment:
The only place where the `Aux` type is used is in the private `instance` method's explicit return type, which is not really needed. I wonder if this is worth it.